go/ql/src/experimental/CWE-347/JWTParsingAlgorithm.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @name JWT Method Check
+ * @description Trusting the Method provided by the incoming JWT token may lead to an algorithim confusion
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 8.0
+ * @precision medium
+ * @id go/jwt-alg-confusion
+ * @tags security
+ *       external/cwe/cwe-347
+ */
+
+import go
+import experimental.frameworks.JWT
+import DataFlow
+
+/**
+ * A parse function that verifies signature and accepts all methods.
+ */
+class SafeJwtParserFunc extends Function {
+  SafeJwtParserFunc() { this.hasQualifiedName(golangJwtModern(), ["Parse", "ParseWithClaims"]) }
+}
+
+/**
+ * A parse method that verifies signature.
+ */
+class SafeJwtParserMethod extends Method {
+  SafeJwtParserMethod() {
+    this.hasQualifiedName(golangJwtModern(), "Parser", ["Parse", "ParseWithClaims"])
+  }
+}
+
+from CallNode c, Function func
+where
+  (
+    c.getTarget() = func and
+    // //Flow from NewParser to Parse (check that this call to Parse does not use a Parser that sets Valid Methods)
+    (
+      func instanceof SafeJwtParserMethod and
+      not exists(CallNode c2, WithValidMethods wvm, NewParser m |
+        c2.getTarget() = m and
+        (
+          c2.getCall().getAnArgument() = wvm.getACall().asExpr() and
```

Comment:
Is it possible to specify which argument?

Reply:
Also I think you could write this as `c2.getArgument(i) = wvm.getACall()` (for some value of `i`), which avoids dropping from data flow nodes to exprs when we don't need to.

Reply:
1. Parse takes a variable number of arguments, all of which are Parse options, so the WithValidMethods could be any argument [relevant docs](https://pkg.go.dev/github.com/golang-jwt/jwt/v5#NewParser). 
2. When I use `[c2.getAnArgument(), c2.getArgument(i)] = wvm.getACall()`, I am not getting any results. 
Example:
```
from CallNode c2, WithValidMethods wvm, NewParser m, int i
where
c2.getTarget() = m and c2.getArgument(i) = wvm.getACall()
select c2, "Here are some NewParser calls that have  WithValidMethods" as argument"
```
returns no results even though the original `c2.getCall().getAnArgument() = wvm.getACall().asExpr()` returns the correct results.

Reply:
Fixed by using getSyntacticArgument